source/grid/grid_generator.cc

```diff
@@ -1910,6 +1910,355 @@ namespace GridGenerator
+                         const double             pad_left,
+                         const double             pad_right,
+                         const Point<2>           new_center,
+                         const types::manifold_id polar_manifold_id,
+                         const types::manifold_id tfi_manifold_id,
+                         const double             L,
+                         const unsigned int       n_slices,
+                         const bool               colorize)
+  {
+    internal::plate_with_a_hole(tria,
+                                inner_radius,
+                                outer_radius,
+                                pad_bottom,
+                                pad_top,
+                                pad_left,
+                                pad_right,
+                                new_center,
+                                polar_manifold_id,
+                                tfi_manifold_id,
+                                L,
+                                n_slices,
+                                colorize);
+
+    // reset material id
+    for (auto &cell : tria.active_cell_iterators())
+      cell->set_material_id(0);
+  }
+
+
+
+  template <>
+  void plate_with_a_hole(Triangulation<3> &       tria,
+                         const double             inner_radius,
+                         const double             outer_radius,
+                         const double             pad_bottom,
+                         const double             pad_top,
+                         const double             pad_left,
+                         const double             pad_right,
+                         const Point<3>           new_center,
+                         const types::manifold_id polar_manifold_id,
+                         const types::manifold_id tfi_manifold_id,
+                         const double             L,
+                         const unsigned int       n_slices,
+                         const bool               colorize)
+  {
+    Triangulation<2> tria_2;
+    internal::plate_with_a_hole(tria_2,
+                                inner_radius,
+                                outer_radius,
+                                pad_bottom,
+                                pad_top,
+                                pad_left,
+                                pad_right,
+                                Point<2>(new_center[0], new_center[1]),
+                                polar_manifold_id,
+                                tfi_manifold_id,
+                                L,
+                                n_slices,
+                                colorize);
+
+    // extrude to 3D
+    extrude_triangulation(tria_2, n_slices, L, tria);
+
+    // shift in Z direction to match specified center
+    GridTools::shift(Point<3>(0, 0, new_center[2] - L / 2.), tria);
+
+    // extrude will keep boundary IDs but can not keep manifolds, set them
+    // below:
```

Comment:
I did not read through the documentation of extrude that carefully, but does it set up boundary ids on the top and bottom of the plate (i.e., the new boundaries that come from extrusion) correctly?

Reply:
> does it set up boundary ids on the top and bottom of the plate

yeap:

>  The boundary indicators of the faces of input are going to be assigned to the corresponding side walls in z direction. The bottom and top get the next two free boundary indicators.